### PR TITLE
config: CNF installation (1) introduce new config format and parser

### PR DESCRIPTION
**Pull request by kosstennbl**
## Description
For supporting multiple deployments in one cnf and improving overall config usability and structure - new config format has to be introduced.
Some existing parameters usage need to be changed for easier adaptation.
This change does not replace the old config format.

## Issues:
Refs: #2121

### CODE REVIEW

File: embedded_files/cnf-testsuite-v2-example.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+deployments:
+  helm_charts:
+    - name: coredns
+      helm_repo_name: stable
+      helm_repo_url: https://cncf.gitlab.io/stable
+      helm_chart_name: coredns
+  helm_directories:
+    - name: envoy
+      helm_directory: ../example-cnfs/envoy/envoy/
+
```
[the remark this answers is not in the document]
Comment (kosstennbl) [reply embedded_files/cnf-testsuite-v2-example.yml]: Done

File: src/tasks/utils/cnf_installation/config.cr [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      def after_initialize
+        unless @dynamic
+          @dynamic = DynamicParameters.new()
+        end
+      end
+    end
+
+    class CommonParameters < ConfigBase
+      @service_name : String | Nil
+      @rolling_update_tag : String | Nil
```
Comment (svteb) [inline src/tasks/utils/cnf_installation/config.cr:32]: Aren't several `rolling_*_tag` missing?

Comment (svteb) [reply src/tasks/utils/cnf_installation/config.cr:32]: Additionally these should be part of container_names, right?

File: src/tasks/utils/cnf_installation/config.cr [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+      def initialize()
+      end
+    end
+
+    class DeploymentsConfig < ConfigBase
+      @helm_charts : Array(HelmChartConfig) = [] of HelmChartConfig
+      @helm_dirs : Array(HelmDirectoryConfig) = [] of HelmDirectoryConfig
+      @manifests : Array(ManifestDirectoryConfig) = [] of ManifestDirectoryConfig
+      # deployments.current and all related functionality should be removed with new installation process.
+      @@current : AnyDeploymentConfig | Nil
```
[the remark this answers is not in the document]
Comment (kosstennbl) [reply src/tasks/utils/cnf_installation/config.cr:58]: We aren't changing installation process yet, and the current one uses something like `config.cnf_congig[:release_name]` or other parameters. Crystal's strict typing isn't much friendly towards accessing them in the old way, and for that, `get_deployment_param` was created. `get_install_method` isn't meant to replace anything, testsuite code should be accessing `config.dynamic.install_method`.

File: src/tasks/utils/cnf_installation/config.cr [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        deployment_found = false
+        {@helm_charts, @helm_dirs, @manifests}.each do |deployment_array|
+          if deployment_array
+            
+            # To be removed with new installation process.
+            if deployment_found
+              raise YAML::Error.new("Multiple deployments are not supported yet")
+            else
+              deployment_found = true
+            end
```
[the remark this answers is not in the document]
Comment (kosstennbl) [reply src/tasks/utils/cnf_installation/config.cr:76]: Makes sense, i didn't wanted to create two similar loops one after another, but that can be solved by private method for multiple deployments check.

File: src/tasks/utils/cnf_installation/config.cr (hunk 2 above)
Comment (kosstennbl) [reply src/tasks/utils/cnf_installation/config.cr:32]: Great point, I was looking around that and found out that these variables are useless in config, but then forgot to remove them.

File: src/tasks/utils/cnf_installation/config.cr (hunk 4 above)
[the remark this answers is not in the document]
Comment (svteb) [reply src/tasks/utils/cnf_installation/config.cr:76]: I was just thinking that something along the lines of `if @helm_charts.size + @helm_dirs.size + @manifests.size > 1 then raise`

File: src/tasks/utils/cnf_installation/config.cr (hunk 2 above)
Comment (kosstennbl) [reply src/tasks/utils/cnf_installation/config.cr]: Removed

**Review by martin-mat — APPROVED** [review]
lgtm

**Review by svteb — APPROVED** [review]
lgtm

**Review by Smitholi67 — APPROVED** [review]
lgtm

**Pull request merged**